File: Java8-JavaFX/src/sevenguis/flightbooker/FlightBooker.java
package sevenguis.flightbooker;

import javafx.application.Application;
import javafx.beans.binding.Bindings;
import javafx.beans.binding.When;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.value.ChangeListener;
import javafx.geometry.Insets;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.ComboBox;
import javafx.scene.control.TextField;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;

import java.time.LocalDate;
import java.util.concurrent.Callable;

import static sevenguis.flightbooker.Util.*;

public class FlightBooker extends Application {

    public void start(Stage stage) {
        ComboBox<String> flightType = new ComboBox<>();
        flightType.getItems().addAll("one-way flight", "return flight");
        flightType.setValue("one-way flight");
        TextField startDate = new TextField(dateToString(LocalDate.now()));
        TextField returnDate = new TextField(dateToString(LocalDate.now()));
        Button book = new Button("Book");

        returnDate.disableProperty().bind(flightType.valueProperty().isEqualTo("one-way flight"));
        startDate.styleProperty().bind(Bindings.createStringBinding(() ->
                isDateString(startDate.getText()) ? "" : "-fx-background-color: lightcoral"
                , startDate.textProperty()));
        returnDate.styleProperty().bind(Bindings.createStringBinding(() ->
                isDateString(returnDate.getText()) ? "" : "-fx-background-color: lightcoral"
                , returnDate.textProperty()));
        book.disableProperty().bind(Bindings.createBooleanBinding(() -> {
            if (flightType.getValue().equals("one-way flight")) {
                return !isDateString(startDate.getText());
            } else {
                return !isDateString(startDate.getText()) ||
                       !isDateString(returnDate.getText()) ||
                       stringToDate(startDate.getText()).compareTo(stringToDate(returnDate.getText())) > 0;
            }
        }, flightType.valueProperty(), startDate.textProperty(), returnDate.textProperty()));

        VBox root = new VBox(10, flightType, startDate, returnDate, book);
        root.setPadding(new Insets(10));

        stage.setScene(new Scene(root));
        stage.setTitle("Flight Booker");
        stage.show();
    }

    public static void main(String[] args) {
        launch(args);
    }
}

File: Scala-ScalaFX/src/main/scala/sevenguis/flightbooker/FlightBooker.scala
package sevenguis.flightbooker

import scalafx.application.JFXApp
import scalafx.application.JFXApp.PrimaryStage
import scalafx.scene.Scene
import scalafx.scene.control.{Button, ComboBox, TextField}
import scalafx.scene.layout.VBox
import scalafx.geometry.Insets
import java.time.format.DateTimeFormatter
import java.time.LocalDate
import scalafx.Includes._
import javafx.beans.value.{ObservableValue}
import javafx.beans.binding.Bindings
import java.util.concurrent.Callable
import java.lang.Boolean

object FlightBooker extends JFXApp {
  val flightType = new ComboBox[String](Seq("one-way flight", "return flight"))
  flightType.value = "one-way flight"
  val startDate = new TextField{text=dateToString(LocalDate.now)}
  val returnDate = new TextField{text=dateToString(LocalDate.now)}
  val book = new Button("Book")

  returnDate.disable <== flightType.value === "one-way flight"
  // Ideally, I'd write it the following way
  //startDate.style <==
  //  when (isDateString(startDate.text)) choose "" otherwise "-fx-background-color: lightcoral"
  startDate.style <== Bindings.createStringBinding(new Callable[String] { override def call(): String =
    if (isDateString(startDate.text.value)) "" else "-fx-background-color: lightcoral"
  }, startDate.text)
  // For comparison, a callback based approach
  returnDate.text.addListener((v: ObservableValue[_ <: String], o: String, n: String) =>
    returnDate.style = if (isDateString(n)) "" else "-fx-background-color: lightcoral")
  book.disable <==  Bindings.createBooleanBinding(new Callable[Boolean] { override def call(): Boolean =
    flightType.value.value match {
      case "one-way flight" => !isDateString(startDate.text.value)
      case "return flight" =>
        !isDateString(startDate.text.value) || !isDateString(returnDate.text.value) ||
        stringToDate(startDate.text.value).compareTo(stringToDate(returnDate.text.value)) > 0
    }
  }, flightType.value, startDate.text, returnDate.text)

  stage = new PrimaryStage {
    title = "FlightBooker"
    scene = new Scene {
      content = new VBox(10) {
        padding = Insets(10)
        content = Seq(flightType, startDate, returnDate, book)
      }
    }
  }

  def dateToString(date: LocalDate) = date.format(DateTimeFormatter.ISO_LOCAL_DATE)
  def stringToDate(string: String) = LocalDate.from(DateTimeFormatter.ISO_LOCAL_DATE.parse(string))
  def isDateString(string: String) =
    try {
      DateTimeFormatter.ISO_LOCAL_DATE.parse(string)
      true
    } catch {
      case e: Exception => false
    }
}
